File: backend/services/notificationService.js
import pool from "../config/db.js";
import { decrypt } from "../utils/encrypt.js";

/**
 * Save notification to database for user to see in app
 */
export const createNotification = async (userId, type, title, message, relatedId = null) => {
  try {
    await pool.query(
      `INSERT INTO notifications (user_id, notification_type, title, message, related_id, created_at)
       VALUES ($1, $2, $3, $4, $5, NOW())`,
      [userId, type, title, message, relatedId]
    );
    console.log(`✅ Notification created for user ${userId}: ${title}`);
  } catch (error) {
    console.error("❌ Error creating notification:", error);
  }
};

/**
 * Get all unread notifications for a user
 */
export const getUserNotifications = async (userId) => {
  try {
    const result = await pool.query(
      `SELECT * FROM notifications 
       WHERE user_id = $1 
       ORDER BY created_at DESC 
       LIMIT 50`,
      [userId]
    );
    return result.rows;
  } catch (error) {
    console.error("❌ Error fetching notifications:", error);
    return [];
  }
};

/**
 * Mark notification as read
 */
export const markNotificationAsRead = async (notificationId, userId) => {
  try {
    await pool.query(
      `UPDATE notifications 
       SET is_read = true 
       WHERE id = $1 AND user_id = $2`,
      [notificationId, userId]
    );
  } catch (error) {
    console.error("❌ Error marking notification as read:", error);
  }
};

/**
 * Check medication reminders (called every minute by scheduler)
 */
export const checkMedicationReminders = async () => {
  try {
    const now = new Date();
    const currentTime = now.toTimeString().substring(0, 5); // "HH:MM" format
    
    console.log(`⏰ Checking medication reminders at ${currentTime}...`);

    // Get all medications with reminders enabled
    const result = await pool.query(
      `SELECT m.id, m.user_id, m.name, m.dosage, m.reminder_times
       FROM medications m
       WHERE m.reminder_enabled = true
       AND m.reminder_times IS NOT NULL
       AND array_length(m.reminder_times, 1) > 0`
    );

    for (const med of result.rows) {
      const reminderTimes = med.reminder_times || [];
      
      // Check if current time matches any reminder time
      if (reminderTimes.includes(currentTime)) {
        const medName = decrypt(med.name);
        const medDosage = decrypt(med.dosage);
        
        await createNotification(
          med.user_id,
          'medication',
          '💊 Medication Reminder',
          `Time to take ${medName} (${medDosage})`,
          med.id
        );
        
        console.log(`✅ Sent medication reminder: ${medName} to user ${med.user_id}`);
      }
    }
  } catch (error) {
    console.error("❌ Error checking medication reminders:", error);
  }
};

/**
 * Check appointment reminders (called every hour by scheduler)
 */
export const checkAppointmentReminders = async () => {
  try {
    const now = new Date();
    console.log(`📅 Checking appointment reminders at ${now.toISOString()}...`);

    // Check 24-hour reminders
    const tomorrow = new Date(now.getTime() + 24 * 60 * 60 * 1000);
    const tomorrowDate = tomorrow.toISOString().split('T')[0];
    const tomorrowHour = tomorrow.getHours();

    const result24h = await pool.query(
      `SELECT a.id, a.user_id, a.title, a.appointment_date, a.appointment_time
       FROM appointments a
       WHERE a.reminder_24h = true
       AND a.notified_24h = false
       AND a.appointment_date = $1
       AND EXTRACT(HOUR FROM a.appointment_time::time) = $2`,
      [tomorrowDate, tomorrowHour]
    );

    for (const apt of result24h.rows) {
      const title = decrypt(apt.title);
      const dateStr = new Date(apt.appointment_date).toLocaleDateString();
      
      await createNotification(
        apt.user_id,
        'appointment',
        '📅 Appointment Tomorrow',
        `${title} on ${dateStr} at ${apt.appointment_time}`,
        apt.id
      );

      // Mark as notified
      await pool.query(
        `UPDATE appointments SET notified_24h = true WHERE id = $1`,
        [apt.id]
      );
      
      console.log(`✅ Sent 24h appointment reminder: ${title} to user ${apt.user_id}`);
    }

    // Check 1-hour reminders
    const oneHourLater = new Date(now.getTime() + 60 * 60 * 1000);
    const oneHourDate = oneHourLater.toISOString().split('T')[0];
    const oneHourTime = oneHourLater.toTimeString().substring(0, 5);

    const result1h = await pool.query(
      `SELECT a.id, a.user_id, a.title, a.appointment_date, a.appointment_time
       FROM appointments a
       WHERE a.reminder_1h = true
       AND a.notified_1h = false
       AND a.appointment_date = $1
       AND a.appointment_time::text LIKE $2`,
      [oneHourDate, oneHourTime + '%']
    );

    for (const apt of result1h.rows) {
      const title = decrypt(apt.title);
      
      await createNotification(
        apt.user_id,
        'appointment',
        '⏰ Appointment in 1 Hour',
        `${title} at ${apt.appointment_time}`,
        apt.id
      );

      // Mark as notified
      await pool.query(
        `UPDATE appointments SET notified_1h = true WHERE id = $1`,
        [apt.id]
      );
      
      console.log(`✅ Sent 1h appointment reminder: ${title} to user ${apt.user_id}`);
    }
  } catch (error) {
    console.error("❌ Error checking appointment reminders:", error);
  }
};

export default {
  createNotification,
  getUserNotifications,
  markNotificationAsRead,
  checkMedicationReminders,
  checkAppointmentReminders
};

File: backend1/services/notification_service.py
"""
Notification service — structured logging, parameterized queries.
"""
from __future__ import annotations

from datetime import datetime, timedelta

from db import query
from utils.encrypt import decrypt
from utils.logger import get_logger

logger = get_logger(__name__)


def create_notification(user_id, notif_type, message, related_id=None, title=None):
    try:
        query(
            """INSERT INTO notifications
               (user_id, type, title, message, related_id, created_at)
               VALUES (%s, %s, %s, %s, %s, NOW())""",
            (user_id, notif_type, title or message, message, related_id),
        )
        logger.info(
            "Notification created",
            extra={"user_id": user_id, "type": notif_type, "related_id": related_id},
        )
    except Exception as error:
        logger.error(
            "Failed to create notification",
            extra={"user_id": user_id, "error": str(error)},
        )


def get_user_notifications(user_id, limit=50, offset=0):
    """Return paginated notifications for user_id."""
    try:
        return query(
            """SELECT id, type, title, message, related_id, is_read, created_at
               FROM notifications
               WHERE user_id = %s
               ORDER BY created_at DESC
               LIMIT %s OFFSET %s""",
            (user_id, limit, offset),
        )
    except Exception as error:
        logger.error("Failed to fetch notifications", extra={"user_id": user_id, "error": str(error)})
        return []


def mark_notification_as_read(notification_id, user_id):
    try:
        query(
            """UPDATE notifications SET is_read = true
               WHERE id = %s AND user_id = %s""",
            (notification_id, user_id),
        )
        logger.info(
            "Notification marked as read",
            extra={"user_id": user_id, "notification_id": notification_id},
        )
    except Exception as error:
        logger.error(
            "Failed to mark notification as read",
            extra={"user_id": user_id, "notification_id": notification_id, "error": str(error)},
        )


def check_medication_reminders():
    try:
        now = datetime.now()
        current_time = now.strftime("%H:%M")
        logger.info("Checking medication reminders", extra={"current_time": current_time})

        rows = query(
            """SELECT m.id, m.user_id, m.name, m.dosage, m.reminder_times
               FROM medications m
               WHERE m.reminder_enabled = true
               AND m.reminder_times IS NOT NULL
               AND array_length(m.reminder_times, 1) > 0"""
        )

        for med in rows:
            reminder_times = med.get("reminder_times") or []
            if current_time in reminder_times:
                med_name = decrypt(med["name"]) if med["name"] else ""
                med_dosage = decrypt(med["dosage"]) if med["dosage"] else ""
                create_notification(
                    med["user_id"],
                    "medication_reminder",
                    f"Time to take {med_name} ({med_dosage})",
                    med["id"],
                )
                logger.info("Medication reminder sent",
                    extra={"user_id": med["user_id"], "med_id": med["id"]})
    except Exception as error:
        logger.error("Error checking medication reminders", extra={"error": str(error)})


def check_appointment_reminders():
    try:
        now = datetime.now()
        logger.info("Checking appointment reminders", extra={"time": now.isoformat()})

        tomorrow = now + timedelta(hours=24)
        tomorrow_date = tomorrow.date()
        tomorrow_hour = tomorrow.hour

        result_24h = query(
            """SELECT a.id, a.user_id, a.title, a.appointment_date, a.appointment_time
               FROM appointments a
               WHERE a.reminder_24h = true
               AND a.notified_24h = false
               AND a.appointment_date = %s
               AND EXTRACT(HOUR FROM a.appointment_time::time) = %s""",
            (tomorrow_date, tomorrow_hour),
        )

        for apt in result_24h:
            title = decrypt(apt["title"]) if apt["title"] else ""
            date_str = apt["appointment_date"].strftime("%m/%d/%Y") if apt.get("appointment_date") else ""
            create_notification(
                apt["user_id"], "appointment_reminder",
                f"{title} on {date_str} at {apt['appointment_time']}", apt["id"],
            )
            query("UPDATE appointments SET notified_24h = true WHERE id = %s", (apt["id"],))
            logger.info("24h reminder sent", extra={"user_id": apt["user_id"], "apt_id": apt["id"]})

        one_hour_later = now + timedelta(hours=1)
        one_hour_date = one_hour_later.date()
        one_hour_time = one_hour_later.strftime("%H:%M")

        result_1h = query(
            """SELECT a.id, a.user_id, a.title, a.appointment_date, a.appointment_time
               FROM appointments a
               WHERE a.reminder_1h = true
               AND a.notified_1h = false
               AND a.appointment_date = %s
               AND a.appointment_time::text LIKE %s""",
            (one_hour_date, one_hour_time + "%"),
        )

        for apt in result_1h:
            title = decrypt(apt["title"]) if apt["title"] else ""
            create_notification(
                apt["user_id"], "appointment_reminder",
                f"{title} at {apt['appointment_time']}", apt["id"],
            )
            query("UPDATE appointments SET notified_1h = true WHERE id = %s", (apt["id"],))
            logger.info("1h reminder sent", extra={"user_id": apt["user_id"], "apt_id": apt["id"]})

    except Exception as error:
        logger.error("Error checking appointment reminders", extra={"error": str(error)})